kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +83,354 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
+    if (existingField.isNullable() && !newField.isNullable()) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Cannot tighten the nullability of existing field %s", existingField.getName()));
+    }
+
+    // Both fields are structs, ensure there's no changes in the individual fields
+    if (existingField.getDataType() instanceof StructType
+        && newField.getDataType() instanceof StructType) {
+      StructType existingStruct = (StructType) existingField.getDataType();
+      Map<String, List<StructField>> existingFieldsByName =
+          existingStruct.fields().stream().collect(Collectors.groupingBy(StructField::getName));
+      StructType newStruct = (StructType) newField.getDataType();
+      for (StructField newNestedField : newStruct.fields()) {
+        List<StructField> existingNestedField = existingFieldsByName.get(newNestedField.getName());
+        // The existing field is guaranteed to be well-defined, there should only be one field with
+        // the name
+        if (existingNestedField != null) {
+          validateNoTypeChange(existingNestedField.get(0), newNestedField);
+        }
+      }
+    } else if (existingField.getDataType() instanceof MapType
+        && newField.getDataType() instanceof MapType) {
+      MapType existingMapType = (MapType) existingField.getDataType();
+      MapType newMapType = (MapType) newField.getDataType();
+
+      validateNoTypeChange(existingMapType.getKeyField(), newMapType.getKeyField());
+      validateNoTypeChange(existingMapType.getValueField(), newMapType.getValueField());
+    } else if (existingField.getDataType() instanceof ArrayType
+        && newField.getDataType() instanceof ArrayType) {
+      ArrayType existingArrayType = (ArrayType) existingField.getDataType();
+      ArrayType newArrayType = (ArrayType) newField.getDataType();
+
+      validateNoTypeChange(existingArrayType.getElementField(), newArrayType.getElementField());
+    } else if (!existingField.getDataType().equivalent(newField.getDataType())) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Cannot change the type of existing field %s from %s to %s",
+              existingField.getName(), existingField.getDataType(), newField.getDataType()));
+    }
+  }
+
+  /** Returns a map from flattened column path to the field in the schema */
+  private static Map<String, StructField> columnPathToField(StructType schema) {
+    List<Tuple2<List<String>, StructField>> columnPathToStructField =
+        filterRecursively(
+            schema,
+            true /* recurseIntoMapOrArrayElements */,
+            false /* stopOnFirstMatch */,
+            sf -> true);
+    Map<String, StructField> columnPathToStructFieldMap = new HashMap<>();
+    for (Tuple2<List<String>, StructField> path : columnPathToStructField) {
+      columnPathToStructFieldMap.put(SchemaUtils.concatWithDot(path._1), path._2);
```

Comment:
for followup: looks like we could get the fieldId map here itself and avoid travelling the other map.